File: app/src/main/java/com/eccentricyan/pohe/presentation/article/ArticleListAdapter.java
package com.eccentricyan.pohe.presentation.article;

import android.databinding.DataBindingUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.eccentricyan.pohe.R;
import com.eccentricyan.pohe.databinding.ArticleBinding;
import com.eccentricyan.pohe.di.component.ActivityComponent;
import com.eccentricyan.pohe.domain.entity.Article;

import java.util.Collections;
import java.util.List;

/**
 * Created by shiyanhui on 2017/04/15.
 */

public class ArticleListAdapter extends RecyclerView.Adapter<ArticleListAdapter.ViewHolder> {
    private ActivityComponent component;
    private List<Article> articles;

    public ArticleListAdapter(ActivityComponent component) {
        this.component = component;
        this.articles = Collections.emptyList();
    }

    public void setArticles(List<Article> articles, boolean refresh) {
        if (refresh || this.articles.size() <= 0) {
            this.articles = articles;
        } else {
            this.articles.addAll(articles);
        }
    }

    @Override
    public ArticleListAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        ArticleBinding binding = DataBindingUtil.inflate(
                LayoutInflater.from(parent.getContext()),
                R.layout.article,
                parent,
                false
        );
        return new ArticleListAdapter.ViewHolder(binding, component);
    }

    @Override
    public void onBindViewHolder(ArticleListAdapter.ViewHolder holder, int position) {
        holder.bind(articles.get(position));
    }

    @Override
    public int getItemCount() {
        return articles.size();
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        final ArticleBinding binding;
        final ActivityComponent component;

        ViewHolder(ArticleBinding binding, ActivityComponent component) {
            super(binding.getRoot());
            this.binding = binding;
            this.component = component;
        }

        void bind(Article article) {
            if (binding.getViewModel() == null) {
                binding.setViewModel(new ArticleViewModel(component, article));
            } else {
                binding.getViewModel().setArticle(article);
            }
            binding.executePendingBindings();
        }
    }
}

File: app/src/main/java/com/eccentricyan/pohe/presentation/article/ArticleListAdapter.kt
package com.eccentricyan.pohe.presentation.article

import android.databinding.DataBindingUtil
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup

import com.eccentricyan.pohe.R
import com.eccentricyan.pohe.databinding.ArticleBinding
import com.eccentricyan.pohe.di.component.ActivityComponent
import com.eccentricyan.pohe.domain.entity.Article

import java.util.Collections

/**
 * Created by shiyanhui on 2017/04/15.
 */

class ArticleListAdapter(private val component: ActivityComponent) : RecyclerView.Adapter<ArticleListAdapter.ViewHolder>() {
    var articles: MutableList<Article>? = null

    init {
        this.articles = ArrayList()
    }

    fun setArticles(articles: MutableList<Article>, refresh: Boolean) {
        if (refresh || this.articles!!.size <= 0) {
            this.articles = articles
        } else {
            this.articles!!.addAll(articles)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = DataBindingUtil.inflate<ArticleBinding>(
                LayoutInflater.from(parent.context),
                R.layout.article,
                parent,
                false
        )
        return ArticleListAdapter.ViewHolder(binding, component)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(articles!![position])
    }

    override fun getItemCount(): Int {
        return articles!!.size
    }

    class ViewHolder(val binding: ArticleBinding, val component: ActivityComponent) : RecyclerView.ViewHolder(binding.root) {

        fun bind(article: Article) {
            if (binding.viewModel == null) {
                binding.viewModel = ArticleViewModel(component, article)
            } else {
                binding.viewModel.setArticle(article)
            }
            binding.executePendingBindings()
        }
    }
}